File: cocos/physics/physx/shapes/physx-shape.ts
/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

/* eslint-disable @typescript-eslint/no-unsafe-return */
import { IVec3Like, Quat, Vec3, geometry } from '../../../core';
import { Collider, RigidBody, PhysicsMaterial, PhysicsSystem } from '../../framework';
import { IBaseShape } from '../../spec/i-physics-shape';
import {
    addReference, getShapeFlags, getShapeMaterials, getShapeWorldBounds, getTempTransform,
    PX, removeReference, _trans,
} from '../physx-adapter';
import { EFilterDataWord3 } from '../physx-enum';
import { PhysXSharedBody } from '../physx-shared-body';
import { PhysXWorld } from '../physx-world';
import { PhysXInstance } from '../physx-instance';

export enum EPhysXShapeType {
    SPHERE,
    BOX,
    CAPSULE,
    CYLINDER,
    CONE,
    PLANE,
    TERRAIN,
    MESH,
}

export class PhysXShape implements IBaseShape {
    private static _MESH_SCALE: any;
    static get MESH_SCALE (): any {
        if (!this._MESH_SCALE) { this._MESH_SCALE = new PX.MeshScale(Vec3.ZERO, Quat.IDENTITY); }
        return this._MESH_SCALE;
    }

    get impl (): any { return this._impl; }
    get collider (): Collider { return this._collider; }
    get attachedRigidBody (): RigidBody | null { return null; }

    private static idCounter = 0;

    readonly id: number;
    readonly type: EPhysXShapeType;

    protected _impl: any = null;
    protected _collider: Collider = null as any;
    protected _flags: any;
    protected _sharedBody!: PhysXSharedBody;
    protected _rotation = new Quat(0, 0, 0, 1);
    protected _index = -1;
    protected _word3 = 0;
    protected _isEnabled = false;

    constructor (type: EPhysXShapeType) {
        this.type = type;
        this.id = PhysXShape.idCounter++;
    }

    initialize (v: Collider): void {
        this._collider = v;
        this._flags = getShapeFlags(v.isTrigger);
        this._sharedBody = (PhysicsSystem.instance.physicsWorld as PhysXWorld).getSharedBody(v.node);
        this._sharedBody.reference = true;
        this.onComponentSet();
        addReference(this, this._impl);
    }

    setIndex (v: number): void {
        this._index = v;
    }

    // virtual
    onComponentSet (): void { }

    // virtual
    updateScale (): void { }

    onLoad (): void {
        this.setMaterial(this._collider.sharedMaterial);
        this.setCenter(this._collider.center);
    }

    onEnable (): void {
        this.addToBody();
        this._isEnabled = true;
        this._sharedBody.enabled = true;
    }

    onDisable (): void {
        this.removeFromBody();
        this._isEnabled = false;
        this._sharedBody.enabled = false;
    }

    onDestroy (): void {
        this._sharedBody.reference = false;
        if (this._impl) {
            removeReference(this, this._impl);
            this._impl.release();
            this._impl = null;
        }
        this._flags = null;
        this._collider = null as any;
    }

    setMaterial (v: PhysicsMaterial | null): void {
        const mat = this.getSharedMaterial(v);
        this._impl.setMaterials(getShapeMaterials(mat));
    }

    protected getSharedMaterial (v: PhysicsMaterial | null): any {
        const v1 = (v == null) ? PhysicsSystem.instance.defaultMaterial : v;
        if (!PX.CACHE_MAT[v1.id]) {
            const physics = PhysXInstance.physics;
            const mat = physics.createMaterial(v1.friction, v1.friction, v1.restitution);
            mat.setFrictionCombineMode(PX.CombineMode.eMULTIPLY);
            mat.setRestitutionCombineMode(PX.CombineMode.eMULTIPLY);
            PX.CACHE_MAT[v1.id] = mat;
            return mat;
        }
        const mat = PX.CACHE_MAT[v1.id];
        mat.setStaticFriction(v1.friction);
        mat.setDynamicFriction(v1.friction);
        mat.setRestitution(v1.restitution);
        return mat;
    }

    setAsTrigger (v: boolean): void {
        if (v) {
            this._impl.setFlag(PX.ShapeFlag.eSIMULATION_SHAPE, !v);
            this._impl.setFlag(PX.ShapeFlag.eTRIGGER_SHAPE, v);
        } else {
            this._impl.setFlag(PX.ShapeFlag.eTRIGGER_SHAPE, v);
            this._impl.setFlag(PX.ShapeFlag.eSIMULATION_SHAPE, !v);
        }
        if (this._index >= 0) {
            this._sharedBody.removeShape(this);
            this._sharedBody.addShape(this);
        }
    }

    setCenter (v: IVec3Like): void {
        const pos = _trans.translation;
        const rot = _trans.rotation;
        Vec3.multiply(pos, v, this._collider.node.worldScale);
        Quat.copy(rot, this._rotation);
        const trans = getTempTransform(pos, rot);
        this._impl.setLocalPose(trans);
    }

    getAABB (v: geometry.AABB): void {
        getShapeWorldBounds(this.impl, this._sharedBody.impl, 1, v);
    }

    getBoundingSphere (v: geometry.Sphere): void {
        geometry.AABB.toBoundingSphere(v, this._collider.worldBounds as geometry.AABB);
    }

    setGroup (v: number): void {
        this._sharedBody.setGroup(v);
    }

    getGroup (): number {
        return this._sharedBody.getGroup();
    }

    addGroup (v: number): void {
        this._sharedBody.addGroup(v);
    }

    removeGroup (v: number): void {
        this._sharedBody.removeGroup(v);
    }

    setMask (v: number): void {
        this._sharedBody.setMask(v);
    }

    getMask (): number {
        return this._sharedBody.getMask();
    }

    addMask (v: number): void {
        this._sharedBody.addMask(v);
    }

    removeMask (v: number): void {
        this._sharedBody.removeMask(v);
    }

    updateFilterData (filterData: any): void {
        this._word3 = EFilterDataWord3.DETECT_CONTACT_CCD;
        if (this._collider.needTriggerEvent) {
            this._word3 |= EFilterDataWord3.DETECT_TRIGGER_EVENT;
        }
        if (this._collider.needCollisionEvent) {
            this._word3 |= EFilterDataWord3.DETECT_CONTACT_EVENT | EFilterDataWord3.DETECT_CONTACT_POINT;
        }
        //filterData.word2 = this.id;//useless
        filterData.word3 = this._word3;
        this.setFilerData(filterData);
    }

    updateEventListener (): void {
        if (this._sharedBody) this.updateFilterData(this._sharedBody.filterData);
    }

    updateByReAdd (): void {
        if (this._isEnabled) {
            this.removeFromBody();
            this.addToBody();
        }
    }

    // virtual
    setFilerData (filterData: any): void {
        this._impl.setQueryFilterData(filterData);
        this._impl.setSimulationFilterData(filterData);
    }

    // virtual
    addToBody (): void {
        this._sharedBody.addShape(this);
    }

    // virtual
    removeFromBody (): void {
        this._sharedBody.removeShape(this);
    }
}

File: native/cocos/physics/physx/shapes/PhysXShape.cpp
/****************************************************************************
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "physics/physx/shapes/PhysXShape.h"
#include "base/std/container/unordered_map.h"
#include "physics/physx/PhysXSharedBody.h"
#include "physics/physx/PhysXUtils.h"
#include "physics/physx/PhysXWorld.h"

namespace cc {
namespace physics {

PhysXShape::PhysXShape() : _mCenter(physx::PxIdentity), _mRotation(physx::PxIdentity) {
    _mObjectID = PhysXWorld::getInstance().addWrapperObject(reinterpret_cast<uintptr_t>(this));
};

void PhysXShape::initialize(Node *node) {
    PhysXWorld &ins = PhysXWorld::getInstance();
    _mSharedBody = ins.getSharedBody(node);
    getSharedBody().reference(true);
    onComponentSet();
    insertToShapeMap();
}

void PhysXShape::onEnable() {
    _mEnabled = true;
    if (_mShape) getSharedBody().addShape(*this);
    getSharedBody().enabled(true);
}

void PhysXShape::onDisable() {
    _mEnabled = false;
    if (_mShape) getSharedBody().removeShape(*this);
    getSharedBody().enabled(false);
}

void PhysXShape::onDestroy() {
    getSharedBody().reference(false);
    eraseFromShapeMap();
    PhysXWorld::getInstance().removeWrapperObject(_mObjectID);
}

void PhysXShape::setMaterial(uint16_t id, float f, float df, float r,
                             uint8_t m0, uint8_t m1) {
    if (!_mShape) return;
    PhysXWorld::getInstance().createMaterial(id, f, df, r, m0, m1);
    auto *mat = reinterpret_cast<physx::PxMaterial *>(PhysXWorld::getInstance().getPXMaterialPtrWithMaterialID(id));
    getShape().setMaterials(&mat, 1);
}

void PhysXShape::setAsTrigger(bool v) {
    if (v) {
        getShape().setFlag(physx::PxShapeFlag::eSIMULATION_SHAPE, !v);
        getShape().setFlag(physx::PxShapeFlag::eTRIGGER_SHAPE, v);
    } else {
        getShape().setFlag(physx::PxShapeFlag::eTRIGGER_SHAPE, v);
        getShape().setFlag(physx::PxShapeFlag::eSIMULATION_SHAPE, !v);
    }
    if (_mEnabled) {
        getSharedBody().removeShape(*this);
        getSharedBody().addShape(*this);
    }
}

void PhysXShape::setCenter(float x, float y, float z) {
    _mCenter = physx::PxVec3{x, y, z};
    updateCenter();
}

uint32_t PhysXShape::getGroup() {
    return getSharedBody().getGroup();
}

void PhysXShape::setGroup(uint32_t g) {
    getSharedBody().setGroup(g);
}

uint32_t PhysXShape::getMask() {
    return getSharedBody().getMask();
}

void PhysXShape::setMask(uint32_t m) {
    getSharedBody().setMask(m);
}

void PhysXShape::updateEventListener(EShapeFilterFlag flag) {
}

geometry::AABB &PhysXShape::getAABB() {
    static IntrusivePtr<geometry::AABB> aabb; // this variable is shared with JS with refcounter
    if (!aabb) {
        aabb = new geometry::AABB;
    }
    if (_mShape) {
        auto bounds = physx::PxShapeExt::getWorldBounds(getShape(), *getSharedBody().getImpl().rigidActor);
        pxSetVec3Ext(aabb->center, (bounds.maximum + bounds.minimum) / 2);
        pxSetVec3Ext(aabb->halfExtents, (bounds.maximum - bounds.minimum) / 2);
    }
    return *aabb;
}

geometry::Sphere &PhysXShape::getBoundingSphere() {
    static IntrusivePtr<geometry::Sphere> sphere; // this variable is shared with JS with refcounter
    if (!sphere) {
        sphere = new geometry::Sphere;
    }
    if (_mShape) sphere->define(getAABB());
    return *sphere;
}

void PhysXShape::updateFilterData(const physx::PxFilterData &data) {
}

void PhysXShape::updateCenter() {
    if (!_mShape) return;
    auto &sb = getSharedBody();
    auto *node = sb.getNode();
    node->updateWorldTransform();
    physx::PxTransform local{_mCenter * node->getWorldScale(), _mRotation};
    getShape().setLocalPose(local);
}

void PhysXShape::insertToShapeMap() {
    if (_mShape) {
        getPxShapeMap().insert(std::pair<uintptr_t, uint32_t>(reinterpret_cast<uintptr_t>(&getShape()), getObjectID()));
    }
}

void PhysXShape::eraseFromShapeMap() {
    if (_mShape) {
        getPxShapeMap().erase(reinterpret_cast<uintptr_t>(&getShape()));
    }
}

} // namespace physics
} // namespace cc
